lang/typescript/src/utils.test.ts

```diff
@@ -0,0 +1,113 @@
+import {AdditionalAddressField} from './regions';
+import {RESERVED_DELIMETER, concatAddressField} from './utils';
+
+describe('RESERVED_DELIMETER', () => {
+  test('is a non-breaking space', () => {
+    expect(RESERVED_DELIMETER).toBe(' ');
+    expect(RESERVED_DELIMETER).toBe('\xA0');
+    expect(RESERVED_DELIMETER).toBe('\u00A0');
+    expect(RESERVED_DELIMETER).not.toBe('\x20');
+    expect(RESERVED_DELIMETER).not.toBe(' ');
+  });
+});
+
+describe('concatAddressField', () => {
+  test('creates a concatenated address field with reserved delimeter', () => {
+    const fieldDefinition: AdditionalAddressField[] = [
+      {key: 'streetNumber', required: false},
+      {key: 'streetName', required: false},
+    ];
+    const fieldValues = {
+      streetNumber: '123',
+      streetName: 'Main',
+    };
+    expect(concatAddressField(fieldDefinition, fieldValues)).toBe(
+      '123\u00A0Main',
+    );
+  });
+
+  test('field definition order matters', () => {
+    const fieldDefinitionNumberFirst: AdditionalAddressField[] = [
+      {key: 'streetNumber', required: false},
+      {key: 'streetName', required: false},
+    ];
+    const fieldDefinitionNameFirst: AdditionalAddressField[] = [
+      {key: 'streetName', required: false},
+      {key: 'streetNumber', required: false},
+    ];
+    const fieldValues = {
+      streetNumber: '123',
+      streetName: 'Main',
+    };
+    expect(concatAddressField(fieldDefinitionNumberFirst, fieldValues)).toBe(
+      '123\u00A0Main',
+    );
+    expect(concatAddressField(fieldDefinitionNameFirst, fieldValues)).toBe(
+      'Main\u00A0123',
+    );
+  });
+
+  // TODO: Do we care about the required field for concatenation or just when we do actual field validation on the UI/Atlas?
```

Comment:
The ruby tests seemed to ignore what was required, seemed like the required field is more for validation, not for the concatenation. Is that right? @DominiqueFlaaa @gabypancu @rochlefebvre

Reply:
Exactly. Systems like Core's checkout negociation and Atlas' own address validation will enforce presence for those fields.  As far as concatenation and splitting goes, all fields are optional.

Reply:
Also cleared up by this change which I'm implementing now: https://github.com/Shopify/worldwide/pull/177